### PR TITLE
Rename debian packages to conform to standards

### PR DESCRIPTION
**Pull request by reubenmiller**
## Proposed changes

Rename debian packages to conform to debian guidelines (underscores (`_`) were renamed to dashes (`-`))

Secondary changes (as a result of the renaming)

* Update install script to support both new and old Debian package names (which prompted other improvements)
  * Support downloading artefacts either with `curl` or `wget` (depending what is available)
  * More streamlined output, but still showing verbose information on errors
  * Add `--dry-run` support to check which artefacts will be downloaded
  * Only install `mosquitto` if it is not already installed and use either `apt-get` or `apk` to install it. This removes the explicit dependency for `apt-get` (but not `dpkg`)
  * Automatically use either `sudo` or `su` if the script was not called with `root` permissions
  * Use `apt-get` instead of `apt` (as `apt` is meant for humans not scripts)
  * Add protection against half-downloaded scripts via curl (inspiration taken from docker install script)
  * On errors print a more helpful reason for the error along with details about the user's setup
* Update uninstall script to check for both new and old Debian package names

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### Example Install

```sh
curl -LSsf https://raw.githubusercontent.com/reubenmiller/thin-edge.io/issue-1552-debian-names/get-thin-edge_io.sh | bash -
```

```sh
Thank you for trying thin-edge.io!

Version argument has not been provided, installing latest: 0.8.1
To install a particular version use this script with the version as an argument.
For example: sudo ./get-thin-edge_io.sh 0.8.1
Installing for architecture arm64
Installing mosquitto as prerequisite for thin-edge.io
debconf: delaying package configuration, since apt-utils is not installed

Downloading tedge...OK
Installing tedge...OK

Downloading tedge_mapper...OK
Installing tedge_mapper...OK

Downloading tedge_agent...OK
Installing tedge_agent...OK

Downloading tedge_apt_plugin...OK
Installing tedge_apt_plugin...OK

Downloading tedge_apama_plugin...OK
Installing tedge_apama_plugin...OK

Downloading c8y_configuration_plugin...OK
Installing c8y_configuration_plugin...OK

Downloading c8y_log_plugin...OK
Installing c8y_log_plugin...OK

Downloading tedge_watchdog...OK
Installing tedge_watchdog...OK

thin-edge.io is now installed on your system!

You can go to our documentation to find next steps: https://github.com/thin-edge/thin-edge.io/blob/main/docs/src/howto-guides/003_registration.md
```

### CODE REVIEW

File: get-thin-edge_io.sh [hunk 1]
```diff
@@ -13,126 +32,306 @@ ARGUMENTS:
 
 OPTIONS:
     --minimal   Install only basic set of components - tedge cli and tedge mappers
+    --dry-run   Don't install anything just let me what it is doing
```
Comment (didier-wenzek) [inline get-thin-edge_io.sh:35]:
```suggestion
    --dry-run   Don't install anything, just let me know what it does
```

File: get-thin-edge_io.sh [hunk 2]
```diff
@@ -13,126 +32,306 @@ ARGUMENTS:
...
+        exit 1
+        ;;
+    esac
+
+
+    if is_dry_run; then
+        echo
+        echo "Dry run complete"
+    # Test if tedge command is there and working
+    elif tedge help >/dev/null; then
```
Comment (didier-wenzek) [inline get-thin-edge_io.sh:283]: Nice idea to double check that things are actually installed!

**Review by didier-wenzek — APPROVED** [review]
Approved

**Comment by didier-wenzek** [comment]
> Consider renaming the cargo project and binary names to use - instead of _ because it is now very confusing for users (see example below). This also has a big impact on maintaining the documentation (and reading!). Alternatively we could just rename the binary (not the cargo project name)

* [Naming conventions for rust crates are unclear](https://rust-lang.github.io/api-guidelines/naming.html) and a source of [discussions](https://users.rust-lang.org/t/package-and-crate-naming-convention/67337) and [discussions](https://github.com/rust-lang/api-guidelines/discussions/29)
* Using `kebab-case` for crate names is okay but these must be referenced using `snake_case` from the rust code. So the burden is pushed elsewhere. That said the renaming request is for binary crates - hence this issue doesn't show.

Hence, I'm okay for the renaming of the binary crates and the associated executables.

**Comment by reubenmiller** [comment]
> > Consider renaming the cargo project and binary names to use - instead of _ because it is now very confusing for users (see example below). This also has a big impact on maintaining the documentation (and reading!). Alternatively we could just rename the binary (not the cargo project name)
> 
> * [Naming conventions for rust crates are unclear](https://rust-lang.github.io/api-guidelines/naming.html) and a source of [discussions](https://users.rust-lang.org/t/package-and-crate-naming-convention/67337) and [discussions](https://github.com/rust-lang/api-guidelines/discussions/29)
> * Using `kebab-case` for crate names is okay but these must be referenced using `snake_case` from the rust code. So the burden is pushed elsewhere. That said the renaming request is for binary crates - hence this issue doesn't show.
> 
> Hence, I'm okay for the renaming of the binary crates and the associated executables.

Ok, then I will proceed with the renaming and see how it comes out.

Though I believe that we should create a symlink to the older binary names .e.g. `tedge_mapper` so that existing installations which are using custom init scripts in `system.toml` don't break.

And then we can put a deprecation notice to remove the symlinks in `1.0.0`

File: get-thin-edge_io.sh (hunk 2 above)
Comment (reubenmiller) [reply get-thin-edge_io.sh:283]: actually this was already there in the existing script, so I can't take credit for that.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The commit log history includes unrelated commits and must be cleared.

**Comment by reubenmiller** [comment]
> The commit log history includes unrelated commits and must be cleared.

Yes I had to merge the changes from the main branch as there conflicts. But yes I will squash before committing.

**Comment by didier-wenzek** [comment]
> > The commit log history includes unrelated commits and must be cleared.
> 
> Yes I had to merge the changes from the main branch as there conflicts. But yes I will squash before committing.

More than a squash, you need an interactive rebase droping/removing the commits that have been added here by accident.

**Comment by reubenmiller** [comment]
I experimented with the package renaming, and the folder names can remain to include underscores `_`, but the renaming of the binary was not so convinient as expected.

Below are the current options are:

**Option 1 (preferred) - just rename `package.name`**

```toml
[package]
name = "tedge-watchdog"
```

**Option 2 - keep `package.name` as is, but set `[[bin]]`**

```toml
[package]
name = "tedge_watchdog"

[[bin]]
name = "tedge-watchdog"
# path is required, otherwise cargo complains about not being able to find the main.rs
path = "src/main.rs"
```

Do you agree with the preferred option 1?
@didier-wenzek, @albinsuresh

**Comment by albinsuresh** [comment]
> Do you agree with the preferred option 1? @didier-wenzek, @albinsuresh

Since option 1 is also limited to just the Cargo files and not renaming every single crate directory, I'm fine with that. Option 2 isn't too bad either as we just have a handful of binary crates which needs that extra `path` key.

**Comment by didier-wenzek** [comment]
I also prefer the 1st option.

> **Option 1 (preferred) - just rename `package.name`**

```toml
[package]
name = "tedge-watchdog"
```

**Comment by reubenmiller** [comment]
@didier-wenzek  Ok now the PR is out of its draft status, I also fixed the commit history

**Review by didier-wenzek — APPROVED** [review]
Approved

**Comment by reubenmiller** [comment]
I'm having some second thoughts about the doc changes since the docs are published as soon as the PR is merged.

Also I would like this branch to be fully tested before it gets merges so it does not break `main`, I have provided a built version to @gligorisaev

**Comment by reubenmiller** [comment]
Ok making the decision to merge the changes without reverting the doc changes. It will introduce some breaking changes into the docs shortly before the next 0.9.0 release (in upcoming days).

**Pull request merged**